code/datums/chat_message.dm

```diff
@@ -138,6 +138,9 @@
 	message.maptext_x = (CHAT_MESSAGE_WIDTH - owner.bound_width) * -0.5
 	message.maptext = complete_text
 
+	if (target.loc == owner) // Special case, holding an atom speaking (pAI, recorder...)
```

Comment:
This looks like it wouldn't handle cases where the item is nested more than 1 level deep, perhaps `is_holder_of` is better?

Reply:
Correct

Reply:
What about a pAI which is in a box which is in a backpack?

Reply:
Did that + the optimisations you suggested (notably, caching the icon)